File: py/serialize-and-deserialize-bst.py
# We did it in 2n bits!

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def __init__(self):
        self.struct, self.data = [], []

    def serialize(self, root):
        """Encodes a tree to a single string.
        :type root: TreeNode
        :rtype: str
        """
        # If root is None, put 0 in structure array and return
        if root is None:
            self.struct.append(0)
            return
      
        # Else place 1 in structure array, val in 'data' array
        # and recur for left and right children
        self.struct.append(1)
        self.data.append(root.val)
        self.serialize(root.left)
        self.serialize(root.right)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        :type data: str
        :rtype: TreeNode
        """
        if len(self.struct) < 1:
            return None

        # Remove one item from structure list 
        b = self.struct[0] 
        self.struct.pop(0) 
          
        # If removed bit is 1 
        if b == 1:  
            val = self.data[0] 
            self.data.pop(0) 
          
            #Create a tree node with removed data 
            root = TreeNode(val) 
      
            #And recur to create left and right subtrees 
            root.left = self.deserialize(self.data); 
            root.right = self.deserialize(self.data); 
            return root 
      
        return None
 
    # Preorder traversal to print
    def preorder(self, root): 
        if root: 
            print(f'val: {root.val}', end=' | ')
                  
            if root.left:
                print(f"left child -->: {root.left.val}", end=' ')
            if root.right:
                print(f"right child --> {root.right.val}", end=' ')
            print()
            self.preorder(root.left)
            self.preorder(root.right)


# Build tree manually for testing
root = TreeNode(10)
root.left = TreeNode(20) 
root.right = TreeNode(30)
root.left.left = TreeNode(40)
root.left.right = TreeNode(50)
root.right.right = TreeNode(70)

print("Given Tree")
codec = Codec()
codec.deserialize(codec.serialize(root))
codec.deserialize(codec.serialize(root))
  
codec.preorder(root)

codec.serialize(root)
print("\nEncoded Tree")
print("Structure List")
  
for i in codec.struct:
    print(i, end=' ')

print("\nDataList")
for value in codec.data:
    print(value, end=' ')
  
newroot = codec.deserialize(codec.data) 
  
print("\n\nPreorder Traversal of decoded tree")
codec.preorder(newroot)
# Your Codec object will be instantiated and called as such:
# codec = Codec()

File: cpp/serialize-and-deserialize-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string order;
        preorderDFS(root, order);
        return order;
    }
    
    inline void preorderDFS(TreeNode* root, string& order) {
        if (!root) return ;
        char buff[4];
        memcpy(buff, &(root->val), sizeof(int));
        for (int i = 0; i < 4; ++i) {
            order.push_back(buff[i]);
        }
        preorderDFS(root->left, order);
        preorderDFS(root->right, order);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return reconstruct(data, pos, INT_MIN, INT_MAX);
    }
    
    inline TreeNode* reconstruct(const string& buffer, int& pos, int minValue, int maxValue) {
        if (pos >= buffer.size()) return NULL; //using pos to check whether buffer ends is better than using char* directly.
        
        int value;
        memcpy(&value, &buffer[pos], sizeof(int));
        if (value < minValue or value > maxValue) return NULL;
        
        TreeNode* node = new TreeNode(value);
        pos += sizeof(int);
        node->left = reconstruct(buffer, pos, minValue, value);
        node->right = reconstruct(buffer, pos, value, maxValue);
        return node;
    }
};


// couldn't quit pull off the python solution in cpp :(
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) {
            return "0";
        }

        encrypt += 1;
        decrypt = (to_string(root->val));
        // struct_list.push_back("1");
        // data_list.push_back(to_string(root->val));
        serialize(root->left);
        serialize(root->right);
        return encrypt;
    }
    

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.size() < 1) {
            return nullptr;
        }

        // string b = static_cast<string>(struct_list.front());
        // struct_list.pop_front();
        int count = 0;
        string b = encrypt[0];
        encrypt.erase(0);

        if (b == "1") {
            // string val = static_cast<string>(data_list.front());
            string val = decrypt[0];
            decrypt.erase(0);

            TreeNode* root = new TreeNode(stoi(val));
            // data_list.pop_front();
            root->left = deserialize(data);
            root->right = deserialize(data);
            return root;
        }
        return nullptr;
    }
private:
    string encrypt;
    string decrypt;
};


// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));